db/keyval/filedb/client.go

```diff
@@ -0,0 +1,422 @@
+//  Copyright (c) 2018 Cisco and/or its affiliates.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at:
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package filedb
+
+import (
+	"bytes"
+	"strings"
+	"sync"
+
+	"github.com/ligato/cn-infra/db/keyval/filedb/database"
+	"github.com/ligato/cn-infra/db/keyval/filedb/decoder"
+	"github.com/ligato/cn-infra/db/keyval/filedb/filesystem"
+
+	"github.com/pkg/errors"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/ligato/cn-infra/datasync"
+	"github.com/ligato/cn-infra/db/keyval"
+	"github.com/ligato/cn-infra/logging"
+)
+
+// Client arranges communication between file system and internal database, and is also responsible for upstream events.
+type Client struct {
+	sync.Mutex
+	log logging.Logger
+
+	// A list of filesystem paths. It may be a specific file, or the whole directory. In case a path is a directory,
+	// all files within are processed. If there are another directories inside, they are skipped.
+	cfgPaths []string
+
+	// Path where the status will be stored. Status will be stored either as json or yaml, determined by file suffix.
+	// If the field is empty or point to a directory, status is not propagated to any file.
+	stPath string
+	// Status reader is chosen according to status file extension.
+	stDc decoder.API
+
+	// Internal database mirrors changes in file system. Since the configuration can be only read, it is up to client
+	// to handle difference between configuration revisions. Every database entry consists from three values:
+	//  - path (where the configuration is written)
+	//  - data key
+	//  - data value
+	// Note: database holds only configuration intended for agent with appropriate prefix
+	db database.FilesSystemDB
+
+	// Filesystem handler, provides methods to work with files/directories
+	fsh filesystem.API
```

Comment:
Fields in structs, especially ones with several fields used across multiple files, should contain more verbose name for better code readability. It took me a few scrolls to figure out some of these fields. Few characters long variable names should mostly be used for local variables.

`fsh` -> `fsHandler`
`stPath` -> `statusPath`
`stDc` -> `statusDecoder`